arviz/data/io_pymc3.py

```diff
@@ -249,44 +287,94 @@ def constant_data_to_xarray(self):
         for name, vals in constant_data_vars.items():
             if hasattr(vals, "get_value"):
                 vals = vals.get_value()
+            # this might be a Deterministic, and must be evaluated
+            elif hasattr(self.model[name], 'eval'):
+                    vals = self.model[name].eval()
             vals = np.atleast_1d(vals)
             val_dims = dims.get(name)
             val_dims, coords = generate_dims_coords(
                 vals.shape, name, dims=val_dims, coords=self.coords
             )
             # filter coords based on the dims
             coords = {key: xr.IndexVariable((key,), data=coords[key]) for key in val_dims}
-            constant_data[name] = xr.DataArray(vals, dims=val_dims, coords=coords)
+            try:
+                constant_data[name] = xr.DataArray(vals, dims=val_dims, coords=coords)
+            except ValueError as e:  # pylint: disable=invalid-name
+                raise ValueError("Error translating constant_data variable %s: %s" % (name, e))
         return xr.Dataset(data_vars=constant_data, attrs=make_attrs(library=self.pymc3))
 
     def to_inference_data(self):
         """Convert all available data to an InferenceData object.
 
-        Note that if groups can not be created (i.e., there is no `trace`, so
+        Note that if groups can not be created (e.g., there is no `trace`, so
         the `posterior` and `sample_stats` can not be extracted), then the InferenceData
         will not have those groups.
         """
-        return InferenceData(
-            **{
-                "posterior": self.posterior_to_xarray(),
-                "sample_stats": self.sample_stats_to_xarray(),
-                "posterior_predictive": self.posterior_predictive_to_xarray(),
-                **self.priors_to_xarray(),
-                "observed_data": self.observed_data_to_xarray(),
-                "constant_data": self.constant_data_to_xarray(),
-            }
-        )
+        id_dict = {
+            "posterior": self.posterior_to_xarray(),
+            "sample_stats": self.sample_stats_to_xarray(),
+            "posterior_predictive": self.posterior_predictive_to_xarray(),
+            "predictions": self.predictions_to_xarray(),
+            **self.priors_to_xarray(),
+            "observed_data": self.observed_data_to_xarray(),
+        }
+        if self.predictions:
+            id_dict["predictions_constant_data"] = self.constant_data_to_xarray()
+        else:
+            id_dict["constant_data"] = self.constant_data_to_xarray()
+        return InferenceData(**id_dict)
 
 
 def from_pymc3(
-    trace=None, *, prior=None, posterior_predictive=None, coords=None, dims=None, model=None
+    trace=None,
+    *,
+    prior=None,
+    posterior_predictive=None,
+    predictions=None,
```

Comment:
I think having a `predictions` argument here will be quite confusing. 

I would only use `predictions` argument in `PyMC3Converter` class. Therefore it will not be visible to users in the `from_pymc3` docs but it will still be available for internal usage in `predictions_from_pymc3` (which would then use the class instead of `from_pymc3`, the code is basically the same in both cases).